### ISSUE TITLE
`getObjectById` and `object.id` interfaces are untyped

### ISSUE THREAD

**Issue by mgajda** [screepers/typed-screeps#92]
Currently all Screeps `id` attributes are implemented as `string`, not `Id<Object>`.

### Expected Behavior
It should be `Id<RoomObject>` or similar type argument, so that they can be saved, and then restored:
`getObjectById(id<RoomObject>) -> RoomObject?`
Main point of using object ids is saving them to memory to keep a hang of any particular object for the future.

### Actual Behavior

Currently all Screeps `id` attributes are `string`. Some objects do not have this property at all.

### Your Environment

- `@types/screeps` version: 2.4.1 commit: ed699e3

**Comment by wtfrank** [comment]
I suppose the advantage of this, if you did explicitly overload getObjectById for every Id<T> specification. would be that you could specify in index.d.ts that the CreepMemory of an attack creep has target_id of type Id<Creep|Structure|ConstructionSite> and then when you do let target = Game.getObjectById(creep.memory.target_id) you just get a union type, instead of having to explicitly cast Game.getObjectById<Creep|Structure|ConstructionSite>() every time you use it.

**Comment by mgajda** [comment]
@wtfrank Especially you normally save objects knowing their type. Is it :+1: ?

**Comment by apemanzilla** [comment]
I just tested with the following code, and it simply doesn't work at all. Both `structure` and `creep` are typed as `{}`.

```typescript
declare interface RoomObject {
    roomName: string;
    roomX: number;
    roomY: number;
}

declare interface Creep extends RoomObject {
    name: string;
}

declare interface Structure extends RoomObject {
    hits: number;
}

declare type Id<T> = string;

declare const creepID: Id<Creep>;
declare const structureID: Id<Structure>;

declare function getObjectByID<T>(id: Id<T>): T;

const creep = getObjectByID(creepID);
const structure = getObjectByID(structureID);
```

**Comment by apemanzilla** [comment]
Actually, the `String` interface had completely slipped my mind - something like this seems to work properly:

```typescript
declare interface RoomObject {
    roomName: string;
    roomX: number;
    roomY: number;
}

declare interface Creep extends RoomObject {
    name: string;
}

declare interface Structure extends RoomObject {
    hits: number;
}

declare interface Id<T> extends String {
	/**
	 * This exists only to introduce constraints on T so that differently
	 * paramterized Id types are not assignable to each other
	 * @deprecated
	 */
	readonly __ignoreme?: T;
}

declare const creepID: Id<Creep>;
declare const structureID: Id<Structure>;

declare function getObjectByID<T>(id: Id<T>): T;

const creep = getObjectByID(creepID); // returns Creep
const structure = getObjectByID(structureID); // returns Structure

// no error: Id<T>.__ignoreme is optional
const creepIDfromString: Id<Creep> = "abc";

// compilation error: Id<Structure> is not assignable to Id<Creep>
const testID: typeof creepID = structureID;
```

A change as drastic as this would likely warrant a major version change, it's very likely that it will break existing code.

**Comment by apemanzilla** [comment]
More specifically, the `id` field for objects could be typed as `Id<this>`, which should allow it to properly work with union types.

**Comment by pyrodogg** [comment]
I've reopnend this issue due to #142 and Postcrafter's suggestion for an alternate implementation in the PR for #139

**Comment by pyrodogg** [comment]
## Current functionality
The current implementation takes the approach outlined above in this issues. 

```TypeScript 
const creepID: Id<Creep> = "123"; // OK
const stringID: string = creepID; // Err Type 'Id<Creep>' is not assignable to type 'string'.ts(2322)
const IDwAssert: string = creepID as string; // OK

const creep1 = Game.getObjectById(creepID); // Type of Creep | null
const creep2 = Game.getObjectById<Creep>(stringID); // Type of Creep | null
const creep3 = Game.getObjectById(stringID); // Type of Unknown
```

## Postcrafter's suggestion
Postgrafter posted post-merge feedback on PR #139 
 with an alternative implementation for typing game object Ids. 

```TypeScript
namespace Tag {
  declare const OpaqueTagSymbol: unique symbol;
  export declare class OpaqueTag<T> {
    private [OpaqueTagSymbol]: T;
  }
}
type Id<T> = string & Tag.OpaqueTag<T>;
```
Changing only this implementation of `Id<T>` in `helpers.ts`, the following outcome is observed. 

```TypeScript
const creepID: Id<Creep> = "123"; // Err Type '"123"' is not assignable to type 'Id<Creep>'. Type '"123"' is not assignable to type 'OpaqueTag<Creep>'.ts(2322)
const stringId: string = creepID; // OK
const IDwAssert: Id<Creep> = "123" as Id<Creep>; // OK

const creep1 = Game.getObjectById(creepID); // Type of Creep
const creep2 = Game.getObjectById<Creep>(stringId); // Err Argument of type 'string' is not assignable to parameter of type 'Id<Creep>'. Type 'string' is not assignable to type 'OpaqueTag<Creep>'.ts(2345)
const creep3 = Game.getObjectById(stringID); // Err Argument of type 'string' is not assignable to parameter of type 'Id<Creep>'. Type 'string' is not assignable to type 'OpaqueTag<Creep>'.ts(2345)
```

I like the concept that type asserting seems more natural when used in a 'narrowing' conversion, but i'm not sure of the impacts to `Game.getObjectById()`, especially that `Game.getObjectById<Creep>(stringID)` is no longer valid.

**Comment by eduter** [comment]
I've looked at my own usage of IDs and `getObjectById` and looks like both approaches would work for me. Although, I agree that it's preferable to require a type assertion when narrowing `string` to `Id<T>`, than the opposite.

**Comment by PostCrafter** [comment]
I think casting `stringId` to `Id<Creep>` should be the proper approach to the `Game.getObjectById` example: `Game.getObjectById(stringId as Id<Creep>);`

**Comment by pyrodogg** [comment]
So it matters more in how strictly we want to restrict methods for getting objects by id than the directionality of required casting itself. The current approach is lesser impact since more uses of `Game getObjectById` are still valid. The alternative means the typing must absolutely occur at the parameter.

In both solutions, `Id<T>` cannot be used as an index signature, so explicit asserting to string type is necessary in these situations.

I think it is more consistent that Id's must always be explicitly asserted to be strings.

If a plain string is used to get a game object, then `unknown` is returned to protect against the untyped parameter, which I think is quite reasonable.

**Comment by PostCrafter** [comment]
> I think it is more consistent that Id's must always be explicitly asserted to be strings.
Not quite sure what you refer the consistency to. Every Id is a string, but not every string is an Id,  therefore explicitly casting strings to Ids makes perfect sense while the inverse does not.

I'm curious why you prefer returning `unknown` on an untyped (imo wrongly typed) parameter over a compilation error.

**Comment by pyrodogg** [comment]
If you're doing something with the unknown object it is also a compile time error.

**Comment by kotarou** [comment]
> Every Id is a string, but not every string is an Id, therefore explicitly casting strings to Ids makes perfect sense while the inverse does not.

This is reasonable behaviour. 

`Game.getObjectById<Type>(stringID)` is the behaviour I want to use overall, especially in situations where I have a object stored by ID that I am not sure of the type of. 

So right now my support would be on the current specification.

What is the specifc use-case or advantage that the second approach offers? 
Also, for consistancys sake, if we're using the more restrictive approach, an arguement could be made for a similar restriction on methods that take a room name as a (currently) string arguement.

**Comment by kotarou** [comment]
Actually, thinking on it a little more...

The unknown type is something I really like here. If I only want to test the existance of something, I don't want to spend friction on typing the call or the type of the ID, I just want to throw the ID in with whatever information I might have.

This allows three results:

- The object exists, and you know what it is (the ID was typed)
- The object esists, but you don't know for sure what it is in-code (Unknown)
- It doesn't exist, or you can't see it.

**Comment by eduter** [comment]
After some testing, I really prefer @PostCrafter's solution.
All examples below were tested replacing `interface Id<T>`'s definition by:

```typescript
declare namespace Tag {
  const OpaqueTagSymbol: unique symbol;
  export class OpaqueTag<T> {
    private [OpaqueTagSymbol]: T;
  }
}
type Id<T> = string & Tag.OpaqueTag<T>;
```

@pyrodogg 
> The alternative means the typing must absolutely occur at the parameter.

The typing must occur when you store the id, which I think is a good thing.

```typescript
// definition
interface BuilderMemory extends CreepMemory {
  target: Id<ConstructionSite>
}

// storing
builder.memory.target = constructionSite.id

// retrieving
const target = Game.getObjectById(builder.memory.target)
```

> In both solutions, Id<T> cannot be used as an index signature, so explicit asserting to string type is necessary in these situations. I think it is more consistent that Id's must always be explicitly asserted to be strings.

I think it's more consistent that Ids, which are strings, never need to be asserted to be strings:

```typescript
const creeps: {[id: string]: Creep} = {}

function storeCreep(creep: Creep) {
  creeps[creep.id] = creep
}

function retrieveCreep(id: Id<Creep>): Creep {
  return creeps[id]
}
```

@kotarou 
> `Game.getObjectById<Type>(stringID)` is the behaviour I want to use overall, especially in situations where I have a object stored by ID that I am not sure of the type of.

If you don't know the type, you can't specify the type parameter. If you do, you can use a type assertion on the parameter.

> The unknown type is something I really like here. If I only want to test the existance of something, I don't want to spend friction on typing the call or the type of the ID, I just want to throw the ID in with whatever information I might have.

```typescript
function isObjectVisible(id: string): boolean {
  return !!Game.getObjectById(id as Id<any>)
}
```

**Comment by pyrodogg** [comment]
I think I'm finally Ok with the Tag solution if `Game getObjectById<T>(id: Id<T> | string): T | null` also includes string as a parameter type as it allows the other syntax options for typing the result. I'll look at pushing a new PR this evening for testing.

**Comment by eduter** [comment]
Why would you use strings instead of `Id<T>`? That sounds like opting out of type safety. Is it for backward compatibility?

**Comment by pyrodogg** [comment]
Mostly because I like the flexibility of also being able to type by different syntaxes. It's just as type safe, would return unknown type if a plain string is provided which is what the original issue was going for.

**Comment by eduter** [comment]
It returns `unkonwn`, which is unusable, so it forces you to use a type assertion, defeating the purpose of this change. If it didn't accept string, you'd have to either use strings and assert them to be Ids, or use Ids all the way. The later being cleaner, easier and safer. `getObjectById` accepting strings encourages the least safe approach, while not accepting it, encourages the safest.

**Comment by pyrodogg** [comment]
Check out the PR, I've implemented the suggested "Tag" based typing for `Id<T>`. 

For backwards compatibility and to allow for alternative syntax for typing the result of `Game.getObjectById()`, `string` typed ids are still allowed. String ids will return `unknown` type, as in the original implementation if the type of the object is not otherwise defined.

Users should be encouraged to use typed ids to get the benefit of automatic type handling but are still allowed to use string Ids if they otherwise assert the type of the object.

**Comment by kotarou** [comment]
Because this is a major release, we don't need to be specifically concerned with backwards compatibility. 

I like keeping the behavior of string IDs returning an unknown type, but I can see the argument that typed-screeps (being more strict than the previous screeps-typescript-declerations repo) should be more restrictive. 

Are there any typescript usecases on which a string is needed without casting to Id<any>? I admit I have not previously though of using Id<any> but it is a pretty obvious and useful output.

**Comment by pyrodogg** [comment]
I've submitted an alternative PR #145 which implements the Tag based implementation and no longer allows string typed Ids to get game objects (ex. `creep = Game.getObjectById<Creep>(stringId)`)

**Comment by kotarou** [comment]
Poll: Which solution do people prefer?

/polls "IDs are generated using  PR #139 without tags" "IDs use tags only" "IDs use tags, but string IDs can be used for unknown type"

**Comment by eduter** [comment]
I think the stricter approach (option 2) is definitely better and, since this is a new major version, breaking changes are to be expected. On the other hand, people who want to use the new stuff (i.e. factories, store, etc) are forced to upgrade. To be nice with them, a fourth alternative would be to not forbid the use of strings as IDs, but deprecate it. Somewhere between options 2 and 3.

```typescript
/**
 * Get an object with the specified unique ID. It may be a game object of any type. Only objects from the rooms which are visible to you can be accessed.
 * @param id The unique identifier.
 * @returns an object instance or null if it cannot be found.
 */
getObjectById<T>(id: Id<T>): T | null;

/**
 * Get an object with the specified unique ID. It may be a game object of any type. Only objects from the rooms which are visible to you can be accessed.
 * @param id The unique identifier.
 * @returns an object instance or null if it cannot be found.
 * @deprecated Use Id<T>, instead of strings, to increase type safety
 */
// tslint:disable-next-line:unified-signatures
getObjectById<T>(id: string): T | null;
```

**Comment by kotarou** [comment]
@eduter That is a wonderful idea.

**Comment by pyrodogg** [comment]
I've updated Option 3 (#144) to incorporate the suggestion from @eduter to mark use of string Ids as deprecated. It has the same effect but the documentation should be clearer.

**Comment by pyrodogg** [comment]
@PostCrafter @eduter @kotarou Can we all agree now that #144 is the most acceptable solution for 3.0? 

It uses the `Tag` method for `Id<T>` and while supporting `string` ids to get game objects, officially marks them as deprecated and returns `unknown` type if not manually asserted.

**Comment by kotarou** [comment]
Yup.

**Comment by pyrodogg** [comment]
I've merged #144 so it's the new "current" method.
